{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS Catalog and Summary Statistics REST API workshop\n",
    "\n",
    "* The following shows basic examples of how to access and parse data from the GWAS Catalog through the REST API. \n",
    "* Although this example is written in Python, any other programming language is equally good.\n",
    "* There are TWO REST APIs, the GWAS Catalog API and the Summary Statistics API.\n",
    "* Examples in other languages will be available soon.\n",
    "\n",
    "\n",
    "### Contents:\n",
    "\n",
    "* **Exercise 1**: fetching data from the API manually, via a browser\n",
    "* **Exercise 2**: fetching data for a list of variants (GWAS Catalog API)\n",
    "* **Exercise 3**: fetching summary statistics data for a genomic region (Summary Stats API)\n",
    "* **Exercise 4**: combining the two APIs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1\n",
    "### _Requests are just URLs_\n",
    "\n",
    "Fetch data of a single study with accession ID [GCST001795](https://www.ebi.ac.uk/gwas/studies/GCST001795) from the GWAS Catalog REST API using a browser.   \n",
    "\n",
    "**Generate the URL:**\n",
    "\n",
    "* API URL: `https://www.ebi.ac.uk/gwas/rest/api`\n",
    "* Endpoint: `studies`\n",
    "* AccessionID: `GCST001795`\n",
    "\n",
    "**URL:**\n",
    "\n",
    "[https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001795](https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001795)\n",
    "\n",
    "Visit the URL in a browser to see the response from the REST API.\n",
    "\n",
    "### Understanding the returned data:\n",
    "\n",
    "* Number of simple key-value pairs eg:\n",
    "\n",
    "```json\n",
    "    \"initialSampleSize\" : \"1,656 Han Chinese ancestry cases, 3,394 Han Chinese ancestry controls\",\n",
    "    \"snpCount\" : 2100739,\n",
    "    \"imputed\" : true,\n",
    "    \"accessionId\" : \"GCST001795\",\n",
    "```\n",
    "\n",
    "* List allowing multiple elements for a key:\n",
    "\n",
    "```json\n",
    "    \"genotypingTechnologies\" : [ {\n",
    "        \"genotypingTechnology\" : \"Genome-wide genotyping array\"\n",
    "    } ],\n",
    "```\n",
    "* List where the values are themselves complex objects eg. ancestries.\n",
    "\n",
    "\n",
    "* The returned data is highly structured, easy to read for computer. \n",
    "* The same information is accessible via the UI.\n",
    "\n",
    "In the following examples we make small scripts in Python to organize this data to make is easy to read for humans."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2\n",
    "\n",
    "### _Retrieve data for a list of variants_\n",
    "\n",
    "1. Fetch the trait and p-value of all associations for multiple rsIDs. \n",
    "2. Organize the data in a table.\n",
    "3. Be careful, might not all rsIDs have associations!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Warning] rs9879858765 is not in the GWAS Catalog!!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>trait</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000e-06</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "      <td>rs142968358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8.000000e-06</td>\n",
       "      <td>lung carcinoma,smoking status measurement</td>\n",
       "      <td>rs62402518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.000000e-07</td>\n",
       "      <td>acute myeloid leukemia</td>\n",
       "      <td>rs12199222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7.000000e-07</td>\n",
       "      <td>body height</td>\n",
       "      <td>rs12199222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>systemic lupus erythematosus</td>\n",
       "      <td>rs7329174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6.000000e-06</td>\n",
       "      <td>systemic lupus erythematosus</td>\n",
       "      <td>rs7329174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8.000000e-09</td>\n",
       "      <td>crohn's disease</td>\n",
       "      <td>rs7329174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3.000000e-06</td>\n",
       "      <td>systemic lupus erythematosus</td>\n",
       "      <td>rs7329174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pvalue                                      trait      variant\n",
       "0  1.000000e-06                   autism spectrum disorder  rs142968358\n",
       "1  8.000000e-06  lung carcinoma,smoking status measurement   rs62402518\n",
       "2  6.000000e-07                     acute myeloid leukemia   rs12199222\n",
       "3  7.000000e-07                                body height   rs12199222\n",
       "4  1.000000e-08               systemic lupus erythematosus    rs7329174\n",
       "5  6.000000e-06               systemic lupus erythematosus    rs7329174\n",
       "6  8.000000e-09                            crohn's disease    rs7329174\n",
       "7  3.000000e-06               systemic lupus erythematosus    rs7329174"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import required packages\n",
    "import requests     # HTTP library - manages data transfer from web resource (e.g. GWAS Catalog)\n",
    "import json         # Handling the json response\n",
    "import pandas as pd # Data analysis library, a bit like R for Python!\n",
    "\n",
    "\n",
    "# API Address:\n",
    "apiUrl = 'https://www.ebi.ac.uk/gwas/rest/api'\n",
    "\n",
    "# List of variants:\n",
    "variants = ['rs142968358', 'rs62402518', 'rs12199222', 'rs7329174', 'rs9879858765']\n",
    "\n",
    "# Store extracted data in this list:\n",
    "extractedData = []\n",
    "\n",
    "# Iterating over all variants:\n",
    "for variant in variants:\n",
    "\n",
    "    # Accessing data for a single variant:\n",
    "    requestUrl = '%s/singleNucleotidePolymorphisms/%s/associations?projection=associationBySnp' %(apiUrl, variant)\n",
    "    response = requests.get(requestUrl, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    \n",
    "    # Testing if rsID exists:\n",
    "    if not response.ok:\n",
    "        print(\"[Warning] %s is not in the GWAS Catalog!!\" % variant)\n",
    "        continue\n",
    "    \n",
    "    # Test if the returned data looks good:\n",
    "    try:\n",
    "        decoded = response.json()\n",
    "    except:\n",
    "        print(\"[Warning] Failed to encode data for %s\" % variant)\n",
    "        continue\n",
    "    \n",
    "    for association in decoded['_embedded']['associations']:\n",
    "        trait = \",\".join([trait['trait'] for trait in association['efoTraits']])\n",
    "        pvalue = association['pvalue']\n",
    "        \n",
    "        extractedData.append({'variant' : variant,\n",
    "                              'trait' : trait,\n",
    "                              'pvalue' : pvalue})\n",
    "        \n",
    "# Format data into a table (data frame):\n",
    "table = pd.DataFrame.from_dict(extractedData)\n",
    "table "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 3\n",
    "### _Summary Statistics API_\n",
    "\n",
    "* Get all associations for chromosome 14 between 1kbp region of 68713000 and 68714000\n",
    "* Filter by p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>efo_trait</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>studyID</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EFO_0004584</td>\n",
       "      <td>7.891000e-11</td>\n",
       "      <td>GCST004599</td>\n",
       "      <td>rs382179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EFO_0004308</td>\n",
       "      <td>5.054000e-12</td>\n",
       "      <td>GCST004610</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>1.189000e-11</td>\n",
       "      <td>GCST004613</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EFO_0007987</td>\n",
       "      <td>4.878000e-11</td>\n",
       "      <td>GCST004614</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>4.591000e-11</td>\n",
       "      <td>GCST004620</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EFO_0007988</td>\n",
       "      <td>2.782000e-10</td>\n",
       "      <td>GCST004626</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>8.026000e-12</td>\n",
       "      <td>GCST004629</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     efo_trait        pvalue     studyID     variant\n",
       "0  EFO_0004584  7.891000e-11  GCST004599    rs382179\n",
       "1  EFO_0004308  5.054000e-12  GCST004610  rs10138752\n",
       "2  EFO_0004833  1.189000e-11  GCST004613  rs10138752\n",
       "3  EFO_0007987  4.878000e-11  GCST004614  rs10138752\n",
       "4  EFO_0004833  4.591000e-11  GCST004620  rs10138752\n",
       "5  EFO_0007988  2.782000e-10  GCST004626  rs10138752\n",
       "6  EFO_0004833  8.026000e-12  GCST004629  rs10138752"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# API Address:\n",
    "apiUrl = 'https://www.ebi.ac.uk/gwas/summary-statistics/api'\n",
    "\n",
    "\n",
    "chromosome = '14'\n",
    "bp_lower = '68713000'\n",
    "bp_upper = '68714000'\n",
    "p_upper = '0.000000001'\n",
    "\n",
    "\n",
    "requestUrl = '%s/chromosomes/%s/associations?bp_lower=%s&bp_upper=%s&p_upper=%s' %(apiUrl, chromosome, bp_lower, bp_upper, p_upper)\n",
    "response = requests.get(requestUrl, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "# The returned response is a \"response\" object, from which we have to extract and parse the information:\n",
    "decoded = response.json()\n",
    "extractedData = []\n",
    "\n",
    "for association in decoded['_embedded']['associations'].values():\n",
    "    trait = \",\".join([trait for trait in association['trait']])\n",
    "    pvalue = association['p_value']\n",
    "    variant = association['variant_id']\n",
    "    studyID = association['study_accession']\n",
    "    \n",
    "    extractedData.append({'variant' : variant,\n",
    "                          'efo_trait' : trait,\n",
    "                          'studyID': studyID,\n",
    "                          'pvalue' : pvalue})    \n",
    "    \n",
    "ssRegionTable = pd.DataFrame.from_dict(extractedData)\n",
    "ssRegionTable \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 4\n",
    "### _Combine the two APIs!_\n",
    "\n",
    "* Get all associations for chromosome 14 between 68713000 and 68714000\n",
    "* Filter by p-value\n",
    "* Add in the pubmed ID and trait name from the study info from the GWAS Catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>efo_trait</th>\n",
       "      <th>pubmedID</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>studyID</th>\n",
       "      <th>traitName</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EFO_0004584</td>\n",
       "      <td>27863252</td>\n",
       "      <td>7.891000e-11</td>\n",
       "      <td>GCST004599</td>\n",
       "      <td>Mean platelet volume</td>\n",
       "      <td>rs382179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EFO_0004308</td>\n",
       "      <td>27863252</td>\n",
       "      <td>5.054000e-12</td>\n",
       "      <td>GCST004610</td>\n",
       "      <td>White blood cell count</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>27863252</td>\n",
       "      <td>1.189000e-11</td>\n",
       "      <td>GCST004613</td>\n",
       "      <td>Sum neutrophil eosinophil counts</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EFO_0007987</td>\n",
       "      <td>27863252</td>\n",
       "      <td>4.878000e-11</td>\n",
       "      <td>GCST004614</td>\n",
       "      <td>Granulocyte count</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>27863252</td>\n",
       "      <td>4.591000e-11</td>\n",
       "      <td>GCST004620</td>\n",
       "      <td>Sum basophil neutrophil counts</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EFO_0007988</td>\n",
       "      <td>27863252</td>\n",
       "      <td>2.782000e-10</td>\n",
       "      <td>GCST004626</td>\n",
       "      <td>Myeloid white cell count</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EFO_0004833</td>\n",
       "      <td>27863252</td>\n",
       "      <td>8.026000e-12</td>\n",
       "      <td>GCST004629</td>\n",
       "      <td>Neutrophil count</td>\n",
       "      <td>rs10138752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     efo_trait  pubmedID        pvalue     studyID  \\\n",
       "0  EFO_0004584  27863252  7.891000e-11  GCST004599   \n",
       "1  EFO_0004308  27863252  5.054000e-12  GCST004610   \n",
       "2  EFO_0004833  27863252  1.189000e-11  GCST004613   \n",
       "3  EFO_0007987  27863252  4.878000e-11  GCST004614   \n",
       "4  EFO_0004833  27863252  4.591000e-11  GCST004620   \n",
       "5  EFO_0007988  27863252  2.782000e-10  GCST004626   \n",
       "6  EFO_0004833  27863252  8.026000e-12  GCST004629   \n",
       "\n",
       "                          traitName     variant  \n",
       "0              Mean platelet volume    rs382179  \n",
       "1            White blood cell count  rs10138752  \n",
       "2  Sum neutrophil eosinophil counts  rs10138752  \n",
       "3                 Granulocyte count  rs10138752  \n",
       "4    Sum basophil neutrophil counts  rs10138752  \n",
       "5          Myeloid white cell count  rs10138752  \n",
       "6                  Neutrophil count  rs10138752  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def getStudy(studyLink):\n",
    "    # Accessing data for a single study:\n",
    "    response = requests.get(studyLink, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    decoded = response.json()\n",
    "    \n",
    "    gwasData = requests.get(decoded['_links']['gwas_catalog']['href'], headers={ \"Content-Type\" : \"application/json\"})\n",
    "    decodedGwasData = gwasData.json()\n",
    "\n",
    "    traitName = decodedGwasData['diseaseTrait']['trait']\n",
    "    pubmedId = decodedGwasData['publicationInfo']['pubmedId']\n",
    "    \n",
    "    return(traitName, pubmedId)\n",
    "\n",
    "\n",
    "# API Address:\n",
    "apiUrl = 'https://www.ebi.ac.uk/gwas/summary-statistics/api'\n",
    "\n",
    "\n",
    "chromosome = '14'\n",
    "bp_lower = '68713000'\n",
    "bp_upper = '68714000'\n",
    "p_upper = '0.000000001'\n",
    "\n",
    "\n",
    "requestUrl = '%s/chromosomes/%s/associations?bp_lower=%s&bp_upper=%s&p_upper=%s' %(apiUrl, chromosome, bp_lower, bp_upper, p_upper)\n",
    "response = requests.get(requestUrl, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "# The returned response is a \"response\" object, from which we have to extract and parse the information:\n",
    "decoded = response.json()\n",
    "extractedData = []\n",
    "\n",
    "for association in decoded['_embedded']['associations'].values():\n",
    "    trait = \",\".join([trait for trait in association['trait']])\n",
    "    pvalue = association['p_value']\n",
    "    variant = association['variant_id']\n",
    "    studyID = association['study_accession']\n",
    "    studyLink = association['_links']['study']['href']\n",
    "    traitName, pubmedId = getStudy(studyLink)\n",
    "    \n",
    "    extractedData.append({'variant' : variant,\n",
    "                          'efo_trait' : trait,\n",
    "                          'studyID': studyID,\n",
    "                          'pvalue' : pvalue,\n",
    "                          'traitName': traitName,\n",
    "                          'pubmedID': pubmedId}) \n",
    "\n",
    "    \n",
    "ssRegionTable = pd.DataFrame.from_dict(extractedData)\n",
    "ssRegionTable \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
