{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS Catalog and Summary Statistics REST API workshop\n",
    "\n",
    "* The following shows basic examples of how to access and parse data from the GWAS Catalog through the REST API. \n",
    "* Although this example is written in Python, any other programming language is equally good.\n",
    "* There are TWO REST APIs, the GWAS Catalog API and the Summary Statistics API.\n",
    "* Examples in other languages will be available soon.\n",
    "\n",
    "\n",
    "### Contents:\n",
    "\n",
    "* **Exercise 1**: fetching data from the API manually, via a browser\n",
    "* **Exercise 2**: fetching data for a list of variants (GWAS Catalog API)\n",
    "* **Exercise 3**: fetching summary statistics data for a genomic region (Summary Stats API)\n",
    "* **Exercise 4**: combining the two APIs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1\n",
    "### _Requests are just URLs_\n",
    "\n",
    "Fetch data of a single study with accession ID [GCST001795](https://www.ebi.ac.uk/gwas/studies/GCST001795) from the GWAS Catalog REST API using a browser.   \n",
    "\n",
    "**Generate the URL:**\n",
    "\n",
    "* API URL: `https://www.ebi.ac.uk/gwas/rest/api`\n",
    "* Endpoint: `studies`\n",
    "* AccessionID: `GCST001795`\n",
    "\n",
    "**URL:**\n",
    "\n",
    "[https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001795](https://www.ebi.ac.uk/gwas/rest/api/studies/GCST001795)\n",
    "\n",
    "Visit the URL in a browser to see the response from the REST API.\n",
    "\n",
    "### Understanding the returned data:\n",
    "\n",
    "* Number of simple key-value pairs eg:\n",
    "\n",
    "```json\n",
    "    \"initialSampleSize\" : \"1,656 Han Chinese ancestry cases, 3,394 Han Chinese ancestry controls\",\n",
    "    \"snpCount\" : 2100739,\n",
    "    \"imputed\" : true,\n",
    "    \"accessionId\" : \"GCST001795\",\n",
    "```\n",
    "\n",
    "* List allowing multiple elements for a key:\n",
    "\n",
    "```json\n",
    "    \"genotypingTechnologies\" : [ {\n",
    "        \"genotypingTechnology\" : \"Genome-wide genotyping array\"\n",
    "    } ],\n",
    "```\n",
    "* List where the values are themselves complex objects eg. ancestries.\n",
    "\n",
    "\n",
    "* The returned data is highly structured, easy to read for computer. \n",
    "* The same information is accessible via the UI.\n",
    "\n",
    "In the following examples we make small scripts in Python to organize this data to make is easy to read for humans."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2\n",
    "\n",
    "### _Retrieve data for a list of variants_\n",
    "\n",
    "1. Fetch the trait and p-value of all associations for multiple rsIDs. \n",
    "2. Organize the data in a table.\n",
    "3. Be careful, might not all rsIDs have associations!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required packages\n",
    "import requests     # HTTP library - manages data transfer from web resource (e.g. GWAS Catalog)\n",
    "import json         # Handling the json response\n",
    "import pandas as pd # Data analysis library, a bit like R for Python!\n",
    "\n",
    "\n",
    "# API Address:\n",
    "apiUrl = 'https://www.ebi.ac.uk/gwas/rest/api'\n",
    "\n",
    "# List of variants:\n",
    "variants = ['rs142968358', 'rs62402518', 'rs12199222', 'rs7329174', 'rs9879858765']\n",
    "\n",
    "# Store extracted data in this list:\n",
    "extractedData = []\n",
    "\n",
    "# Iterating over all variants:\n",
    "for variant in variants:\n",
    "\n",
    "    # Accessing data for a single variant:\n",
    "    requestUrl = '%s/singleNucleotidePolymorphisms/%s/associations?projection=associationBySnp' %(apiUrl, variant)\n",
    "    response = requests.get(requestUrl, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    \n",
    "    # Testing if rsID exists:\n",
    "    if not response.ok:\n",
    "        print(\"[Warning] %s is not in the GWAS Catalog!!\" % variant)\n",
    "        continue\n",
    "    \n",
    "    # Test if the returned data looks good:\n",
    "    try:\n",
    "        decoded = response.json()\n",
    "    except:\n",
    "        print(\"[Warning] Failed to encode data for %s\" % variant)\n",
    "        continue\n",
    "    \n",
    "    for association in decoded['_embedded']['associations']:\n",
    "        trait = \",\".join([trait['trait'] for trait in association['efoTraits']])\n",
    "        pvalue = association['pvalue']\n",
    "        \n",
    "        extractedData.append({'variant' : variant,\n",
    "                              'trait' : trait,\n",
    "                              'pvalue' : pvalue})\n",
    "        \n",
    "# Format data into a table (data frame):\n",
    "table = pd.DataFrame.from_dict(extractedData)\n",
    "table "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 3\n",
    "### _Summary Statistics API_\n",
    "\n",
    "* Get all associations for type II diabetes mellitus: EFO_0001360\n",
    "* Filter by p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# API Address:\n",
    "apiUrl = 'https://www.ebi.ac.uk/gwas/summary-statistics/api'\n",
    "\n",
    "\n",
    "trait = \"EFO_0001360\"\n",
    "p_upper = \"0.000000001\"\n",
    "\n",
    "\n",
    "requestUrl = '%s/traits/%s/associations?p_upper=%s&size=10' %(apiUrl, trait, p_upper)\n",
    "response = requests.get(requestUrl, headers={ \"Content-Type\" : \"application/json\"})\n",
    "\n",
    "# The returned response is a \"response\" object, from which we have to extract and parse the information:\n",
    "decoded = response.json()\n",
    "extractedData = []\n",
    "\n",
    "for association in decoded['_embedded']['associations'].values():\n",
    "    pvalue = association['p_value']\n",
    "    variant = association['variant_id']\n",
    "    studyID = association['study_accession']\n",
    "    \n",
    "    extractedData.append({'variant' : variant,\n",
    "                          'studyID': studyID,\n",
    "                          'pvalue' : pvalue})    \n",
    "    \n",
    "ssTable = pd.DataFrame.from_dict(extractedData)\n",
    "ssTable \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 4\n",
    "### _Combine the two APIs!_\n",
    "\n",
    "* Get all associations for type II diabetes mellitus: EFO_0001360\n",
    "* Filter by p-value\n",
    "* Add in the pubmed ID and trait name from the study info from the GWAS Catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getStudy(studyLink):\n",
    "    # Accessing data for a single study:\n",
    "    response = requests.get(studyLink, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    decoded = response.json()\n",
    "    \n",
    "    gwasData = requests.get(decoded['_links']['gwas_catalog']['href'], headers={ \"Content-Type\" : \"application/json\"})\n",
    "    decodedGwasData = gwasData.json()\n",
    "\n",
    "    traitName = decodedGwasData['diseaseTrait']['trait']\n",
    "    pubmedId = decodedGwasData['publicationInfo']['pubmedId']\n",
    "    \n",
    "    return(traitName, pubmedId)\n",
    "\n",
    "\n",
    "extractedData = []\n",
    "\n",
    "for association in decoded['_embedded']['associations'].values():\n",
    "    pvalue = association['p_value']\n",
    "    variant = association['variant_id']\n",
    "    studyID = association['study_accession']\n",
    "    studyLink = association['_links']['study']['href']\n",
    "    traitName, pubmedId = getStudy(studyLink)\n",
    "    \n",
    "    extractedData.append({'variant' : variant,\n",
    "                          'studyID': studyID,\n",
    "                          'pvalue' : pvalue,\n",
    "                          'traitName': traitName,\n",
    "                          'pubmedID': pubmedId}) \n",
    "\n",
    "    \n",
    "ssWithGWASTable = pd.DataFrame.from_dict(extractedData)\n",
    "ssWithGWASTable\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
